%----------------------------------------------------------------------------------------
%	MODULE INFORMATION
%----------------------------------------------------------------------------------------

% Define the top matter
\setModuleTitle{Single Nucleotica Variant Call and Annotation}
\setModuleAuthors{%
  Matt Field \mailto{}\\
  Dan Andrews \mailto{}\\
  Velimir Gayevskiy \mailto{v.gayevskiy@garvan.org.au} \\
}

\setModuleContributions{%
  Gayle Phillip \mailto{xxx@xxx} \\
  Sonika Tyagi \mailto{yyy@yyy}% 
}

%----------------------------------------------------------------------------------------
%	MODULE TITLE PAGE
%----------------------------------------------------------------------------------------

\chapter{\moduleTitle}

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	LEARNING OUTCOMES
%----------------------------------------------------------------------------------------

\section{Key Learning Outcomes}

After completing this practical the trainee should be able to:

\begin{itemize}
  \item LO1 
  \item LO2 
  \item LO3
\end{itemize}

%----------------------------------------------------------------------------------------
%	MODULE RESOURCES
%----------------------------------------------------------------------------------------

\section{Resources You'll be Using}

\subsection{Tools Used}

\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=1cm]
  \item[tool name] \hfill\\
    \url{tool_url/}
  \item[IGV] \hfill\\
    \url{http://www.broadinstitute.org/igv/}
\end{description}

%------------------------------------------------

\subsection{Sources of Data}

\url{http://sra.dnanexus.com/studies/ERP001071}\\
\url{http://www.ncbi.nlm.nih.gov/pubmed/22194472}

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	INTRODUCTION
%----------------------------------------------------------------------------------------

\section{Introduction}

The goal of this hands-on session is to 

To ensure reasonable analysis times, we will perform the analysis on a 

In the second part of the tutorial we will also be using IGV to visualise and manually inspect 

%----------------------------------------------------------------------------------------
%	THE ENVIRONMENT
%----------------------------------------------------------------------------------------

\section{Prepare the Environment}

We will use a dataset derived from whole genome sequencing of a 33-yr-old lung adenocarcinoma patient, who is a never-smoker and has no familial cancer history. 

The data files are contained in the subdirectory called \texttt{data} and are the following:

\begin{description}[style=multiline,labelindent=1.5cm,align=left,leftmargin=2.5cm]
  \item[\texttt{SM\_Blood.merged.mrkdup.realn.chr4\_1st50Mb.bam} and \texttt{SM\_Blood.merged.mrkdup.realn.chr4\_1st50Mb.bam.bai}] \hfill\\
  \item[\texttt{SM\_liverMets.merged.mrkdup.realn.chr4\_1st50Mb.bam} and \texttt{SM\_liverMets.merged.mrkdup.realn.chr4\_1st50Mb.bam}] \hfill\\
  These files are based on subsetting the whole genomes derived from blood and liver metastases to the first 40Mb of chromosome 4. This will allow our analyses to run in a sufficient time during the workshop, but it's worth being aware that these are just 1.25\% of the genome which highlights the length of time and resourced required to perform cancer genomics on full genomes!
\end{description}

\begin{steps}
Open the Terminal and go to the \texttt{SNV} working directory:
\begin{lstlisting}
cd ~/SNV/
\end{lstlisting}
\end{steps}

%\reversemarginpar\marginpar{\vskip+0em\hfill\includegraphics[height=1cm]{graphics/warning.png}}
%\textcolor{red}{
\begin{warning}
  All commands entered into the terminal for this tutorial should be from within the
  \textbf{\texttt{$\sim$/SNV}} directory.
\end{warning}

\begin{steps}
Check that the \texttt{data} directory contains the above-mentioned files by typing:
\begin{lstlisting}
ls data
\end{lstlisting}
\end{steps}

%----------------------------------------------------------------------------------------
%	BAM MANIPULATION
%----------------------------------------------------------------------------------------


\section{BAM Pre-Processing}

text here..

\subsection{Step 1: Pre-Processing}

\begin{steps}
\begin{lstlisting}
command goes here
\end{lstlisting}
\end{steps}

\begin{note}
Explanation of parameters
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
 \item[-n] the normal BAM
 \item[-t] the tumour BAM
 \item[--fasta] the reference genome used for mapping (b37 from GATK here)
 \item[-gc] GC content as windows through the genome (pre-generated and downloadable from the Sequenza website)
\end{description}
\end{note}

This will take approximately 20 minutes to run...

\begin{information}
You can look at the first few lines of the output in the file \texttt{stage1-output.large.seqz.gz} with:
 
\begin{lstlisting}
command goes here
\end{lstlisting}
\end{information}

%------------------------------------------------

\subsection{Step 2: whatever..}

\begin{steps}
\begin{lstlisting}
command goes here
\end{lstlisting}
\end{steps}

%\reversemarginpar\marginpar{\vskip+0em\hfill\includegraphics[height=1cm]{graphics/notes.png}}
\begin{note}
Explanation of parameters
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
 \item[-w] the window size (50 for exomes, 200 for genomes)
 \item[-s] the large seqz file generated in the first step
\end{description}
\end{note}

This step should take approximately X minutes to complete.

%------------------------------------------------

\subsection{Step 3: something}

\begin{steps}
Open the R terminal
\begin{lstlisting}
command
\end{lstlisting}
\end{steps}

commands for R

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	VISUALISATION
%----------------------------------------------------------------------------------------

\section{for example Visualisation}

view PDFs

questions/interpratation

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	NEXT SUBSECTION
%----------------------------------------------------------------------------------------

\section{Variant calls for example}

blurb about what we are going to do next. And a short description of the tools used and how to access it.

\begin{steps}
Step by step instruction to open the files 

\begin{lstlisting}
command
\end{lstlisting}
\end{steps}

text

\begin{note}
Please note that the output files you are creating are saved in your present working directory. If you are not sure where you are in the file system try typing \texttt{pwd} on your command prompt to find out.
\end{note}

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	REFERENCES
%----------------------------------------------------------------------------------------

\section{References}

%TODO Change to using BiBTeX
\begin{enumerate}
  \item Ju YS1, Lee WC, Shin JY, Lee S, Bleazard T, Won JK, Kim YT, Kim JI, Kang JH, Seo JS. A transforming KIF5B and RET gene fusion in lung adenocarcinoma revealed from whole-genome and transcriptome sequencing. Genome Res. 2012 Mar;22(3):436-45. 
\end{enumerate}
